Jesmon Cards

S-E-I-G-E opened the pull request:
Added scripts for BT20-17, 19, and 21.
Added token data for BT20-17 to CardEffectCommons and ContinuousController.

TODO:
Finish coroutines for BT20-19 and 21, Add script for BT20-17-token

S-E-I-G-E committed:
Added scripts for BT20-17, 19, and 21.
Added token data for BT20-17 to CardEffectCommons and ContinuousController.

TODO:
Finish coroutines for BT20-19 and 21, Add script for BT20-17-token

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_017.cs:
```diff
@@ -0,0 +1,207 @@
+using System.Collections.Generic;
+using System.Linq;
+
+namespace DCGO.CardEffects
```
namespace missing .BT20

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_017.cs:
```diff
@@ -0,0 +1,207 @@
+        {
+            List<ICardEffect> cardEffects = new List<ICardEffect>();
+
+            #region OnPlay WhenDigivovling
```
On play /when digivolving effects MUST be separated (thank venusmon)

S-E-I-G-E replied:
listed in order under same timing check, scoped with curly brace scopes

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_017.cs:
```diff
@@ -0,0 +1,207 @@
+
+                string EffectDiscription()
+                {
+                    return "Play 1 [Atho, Ren� & Por] Token. (Digimon/White/6000 DP/<Reboot> <Blocker> <Decoy Red/Black>)";
```
follow the same as PlayVoleeZerdrucken it had special accent characters in name as well. I cant recall if they were used or not, just copy what it is doing.

Use the FULL effect description here: "[On Play] [When Digivolving] You may play 1 [Atho, Rebe & Por] Token. (Digimon/White/6000 DP/<Reboot> <Blocker> <Decoy<Red>/<Black>>)" taking away on play/When digivolving respectively when you split them.

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_017.cs:
```diff
@@ -0,0 +1,207 @@
+
+                bool CanActivateCondition(Hashtable hashtable)
+                {
+                    if (CardEffectCommons.IsExistOnBattleArea(card))
```
use IsExistOnBattleAreaDigimon instead

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_017.cs:
```diff
@@ -0,0 +1,207 @@
+
+                string EffectDiscription ()
+                {
+                    return "When this Digimon would digivolve or leave the battle area, delete"
```
needs to be the full effect text

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_017.cs:
```diff
@@ -0,0 +1,207 @@
+            if (timing == EffectTiming.OnEnterFieldAnyone)
+            {
+                ActivateClass activateClass = new ActivateClass();
+                activateClass.SetUpICardEffect("Play token", CanUseCondition, card);
```
"play token" is incorrect shorthand text

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_017.cs:
```diff
@@ -0,0 +1,207 @@
+            {
+                ActivateClass activateClass = new ActivateClass();
+                activateClass.SetUpICardEffect("Play token", CanUseCondition, card);
+                activateClass.SetUpActivateClass(null, ActivateCoroutine, -1, true, EffectDiscription());
```
this effect should have activation condition, also should be marked as once per turn. And a SetHashString should be used because its once per turn

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_017.cs:
```diff
@@ -0,0 +1,207 @@
+
+                bool CanUseCondition (Hashtable hashtable)
+                {
+                    if (CardEffectCommons.IsExistOnBattleArea(card))
```
use IsExistOnBattleAreaDigimon instead

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_017.cs:
```diff
@@ -0,0 +1,207 @@
+                {
+                    if (CardEffectCommons.IsExistOnBattleArea(card))
+                    {
+                        if (!CardEffectCommons.IsOpponentTurn(card))
```
use IsOwnerTurn instead

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_017.cs:
```diff
@@ -0,0 +1,207 @@
+                    {
+                        if (!CardEffectCommons.IsOpponentTurn(card))
+                        {
+                            return true;
```
needs to also check if a permanent meeting the condition of being your digimon was played

S-E-I-G-E replied:
also not 100% on this one. needs another look

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_017.cs:
```diff
@@ -0,0 +1,207 @@
+                {
+                    if (CardEffectCommons.IsPermanentExistsOnOpponentBattleAreaDigimon(permanent, card))
+                    {
+                        if (permanent.CanBeRemoved() && permanent.DP <= 8000)
```
need to use  card.Owner.MaxDP_DeleteEffect whenever deleting based on DP value

Michael8818 replied:
remove the CanBeRemoved check, it is legal for you to target a digimon for deletion even though you can't delete it

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_017.cs:
```diff
@@ -0,0 +1,207 @@
+                            if (card.Owner.Enemy.GetBattleAreaDigimons().Count((enemyDigimon) => permanent.CanAttackTargetDigimon(enemyDigimon, activateClass)) >= 1)
+                            {
+                                return true;
+                            }
```
not needed, card only says the digimon may attack it doesn't restrict you to only attack digimon

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_017.cs:
```diff
@@ -0,0 +1,207 @@
+                            canTargetCondition_ByPreSelecetedList: null,
+                            canEndSelectCondition: null,
+                            maxCount: maxCount,
+                            canNoSelect: true,
```
this is a mandatory deletion, canNoSelect should be false

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_017.cs:
```diff
@@ -0,0 +1,207 @@
+                                    selectAttackEffect.SetUp(
+                                        attacker: selectedPermanent,
+                                        canAttackPlayerCondition: () => false,
+                                        defenderCondition: (permanent) => true,
```
this effect doesn't restrict the attack target, canAttackPlayerCondition should be true

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_019.cs:
```diff
@@ -0,0 +1,318 @@
+using System.Collections.Generic;
+using System.Linq;
+
+namespace DCGO.CardEffects
```
namespace missing .BT20

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_019.cs:
```diff
@@ -0,0 +1,318 @@
+        {
+            List<ICardEffect> cardEffects = new List<ICardEffect>();
+
+            if (timing == EffectTiming.None)
```
incorrect timing for when digivolving effect. Also place in a region for easier readability

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_019.cs:
```diff
@@ -0,0 +1,318 @@
+            {
+                ActivateClass activateClass = new ActivateClass();
+                activateClass.SetUpICardEffect("Become unaffected by opponents effects, then can attack", CanUseCondition, card);
+                activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, -1, true, EffectDiscription());
```
since this isn't a once per turn effect you may want to not use, should make it mandatory (false)

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_019.cs:
```diff
@@ -0,0 +1,318 @@
+                {
+                    if (CardEffectCommons.CanTriggerWhenDigivolving(hashtable, card))
+                    {
+                        return CardEffectCommons.IsExistOnBattleArea(card);
```
use IsExistOnBattleAreaDigimon

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_019.cs:
```diff
@@ -0,0 +1,318 @@
+                {
+                    if (CardEffectCommons.CanTriggerWhenDigivolving(hashtable, card))
+                    {
+                        if (CardEffectCommons.IsExistOnBattleArea(card))
```
use IsExistOnBattleAreaDigimon instead

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_019.cs:
```diff
@@ -0,0 +1,318 @@
+                    return false;
+                }
+
+                bool CanActivateCondition(Hashtable hashtable)
```
the only condition for its activation is if it exists in battle area

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_019.cs:
```diff
@@ -0,0 +1,318 @@
+                                    canNoSelect: true,
+                                    canEndNotMax: false,
+                                    selectPermanentCoroutine: null,
+                                    afterSelectPermanentCoroutine: AfterSelectPermanentCoroutine,
```
no need to do afterSelectPermanentCoroutine. use selectPermanentCoroutine instead here

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_019.cs:
```diff
@@ -0,0 +1,318 @@
+                                    yield return null;
+                                }
+
+                                foreach (Permanent permanent in selectedPermanents)
```
when you switch to use selectPermanentCoroutine looping through a list will not be needed.

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_019.cs:
```diff
@@ -0,0 +1,318 @@
+                                                defenderCondition: (permanent) => true,
+                                                cardEffect: activateClass);
+
+                                            yield return ContinuousController.instance.StartCoroutine(selectAttackEffect.Activate());
```
this should be granting immunity, not attacking

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_019.cs:
```diff
@@ -0,0 +1,318 @@
+                                    {
+                                        if (permanent.CanAttack(activateClass))
+                                        {
+                                            if (card.Owner.Enemy.GetBattleAreaDigimons().Count((enemyDigimon) => permanent.CanAttackTargetDigimon(enemyDigimon, activateClass)) >= 1)
```
effect doesn't restrict the attack target, not needed

Michael8818 replied:
Need to remove this, the effect doesn't restrict attacking opponents digimon. So no need to make sure the opponent HAS digimon to attack.

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_019.cs:
```diff
@@ -0,0 +1,318 @@
+                                {
+                                    selectedPermanents = permanents;
+                                    yield return null;
+                                }
```
same as above no need for the afterSelectPermanent List, use the other one

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_019.cs:
```diff
@@ -0,0 +1,318 @@
+
+                                            selectAttackEffect.SetUp(
+                                                attacker: selectedPermanent,
+                                                canAttackPlayerCondition: () => false,
```
effect doesn't restrict attack target, should be true

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_019.cs:
```diff
@@ -0,0 +1,318 @@
+
+            #region Alliance
+
+            #endregion
```
remove

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_019.cs:
```diff
@@ -0,0 +1,318 @@
+                    yield return null;
+                }
+            }
+            #endregion
```
remove

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_019.cs:
```diff
@@ -0,0 +1,318 @@
+            #endregion
+
+            #region Your Turn
+            if (timing == EffectTiming.OnStartTurn)
```
this will technically have the none timing

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_019.cs:
```diff
@@ -0,0 +1,318 @@
+
+                string EffectDiscription()
+                {
+                    return "[When Digivolving] if [Jesmon]/[X Antibody] is in this Digimon's digivoulution cards, for the turn, 1 of your Digimon isn't affected by you opponent's effects. THne, 1 of your Digimon may attack.";
```
fix text issue with word Then

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_019.cs:
```diff
@@ -0,0 +1,318 @@
+
+                string EffectDiscription ()
+                {
+                    return "All of you DIgimon with [Sistermon] in their names or the [Royal Knight] trait gain <Piercing> and can also attack unsuspended Digimon.";
```
fix DIgimon to Digimon

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_019.cs:
```diff
@@ -0,0 +1,318 @@
+            {
+
+                CanAttackTargetDefendingPermanentClass canAttackTargetDefendingPermanentClass = new CanAttackTargetDefendingPermanentClass();
+                canAttackTargetDefendingPermanentClass.SetUpICardEffect("Gain Piercing and can attack unsuspended Digimon", CanUseCondition, card);
```
only part here should be can attack unsuspended digimon

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_019.cs:
```diff
@@ -0,0 +1,318 @@
+
+                bool CanUseCondition (Hashtable hashtable)
+                {
+                    if (CardEffectCommons.IsExistOnBattleArea(card))
```
use IsExistOnBattleAreaDigimon instead

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_019.cs:
```diff
@@ -0,0 +1,318 @@
+            #endregion
+
+            #region Your Turn
+            if (timing == EffectTiming.OnStartTurn)
```
effect currently doesn't add piercing

S-E-I-G-E replied:
what timing should this effect have? at some point I changed it to None timing and I cannot remember why

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_019.cs:
```diff
@@ -0,0 +1,318 @@
+            }
+            #endregion
+
+            #region Your Turn
```
check out BT20-059 for an example of giving the whole board specific effects if they meet a condition

S-E-I-G-E replied:
I think I got this fixed, but I am not 100%

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_021.cs:
```diff
@@ -0,0 +1,86 @@
+using System.Collections;
+using System.Collections.Generic;
+
+namespace DCGO.CardEffects
```
namespace missing .BT20

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_021.cs:
```diff
@@ -0,0 +1,86 @@
+            }
+            #endregion
+
+            #region OnPlay - WhenDigivolve - When Attacking
```
These must be seperated

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_021.cs:
```diff
@@ -0,0 +1,86 @@
+            {
+                ActivateClass activate_class = new ActivateClass();
+                activate_class.SetUpICardEffect ("Select 1 card, delete 1 card", CanUseCondition, card);
+                activate_class.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, 1, false, EffectDescription());
```
once per turn requires a SetHashString, for all 3 separated effects here the hashstring will be the same.

Michael8818 replied:
Make this optional, only way to not use up the once per turn is to not trigger the effect at all

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_021.cs:
both effects are missing, first On play/when digivolving/when attacking is partially there. Second when attacking is missing entirely

Michael8818 commented:
BT20-008, 013, 014 are missing from the PR. They are apart of the jesmon line. I understand the confusion though. the react only mentions the jesmon/gx but really it is referring to the archetype cards for jesmon. so Huckmon, Baohuck, etc

S-E-I-G-E committed:
Applied feedback

Completed coroutines for BT20 19 and 21

Added Optional parameter to CardEffectCommons.GainPierce to specify if the effect is inherited.

S-E-I-G-E committed:
Added BT20_008, 013 and 014 scripts

008, and 013 scripts are done, finishing up 014

S-E-I-G-E committed:
Finished BT20-008, 013, 014.
Added BT20-17-token

Michael8818 commented on Assets/Scripts/CardEffectCommons/KeyWordEffects/Pierce.cs:
revert these changes it isn't the right way.

Michael8818 commented on Assets/CardEffect/BT20/White/BT20_17_token.cs:
```diff
@@ -0,0 +1,62 @@
+using System.Collections;
+using System.Collections.Generic;
+
+namespace DCGO.CardEffects.BT20
```
namespace should be DCGO.CardEffects.Token instead

Michael8818 commented on Assets/CardEffect/BT20/White/BT20_17_token.cs:
```diff
@@ -0,0 +1,62 @@
+                    return false;
+                }
+
+                cardEffects.Add(CardEffectFactory.DecoySelfEffect(isInheritedEffect: false, card: card, condition: null, permanentCondition: CanSelectDecoyPermanentCondition, effectName: "Decoy ([Black])", effectDiscription: DecoyDiscription()));
```
wrong effect name here

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_021.cs:
```diff
@@ -0,0 +1,742 @@
+                #region OnPlay
+                ActivateClass activate_class = new ActivateClass();
+                activate_class.SetUpICardEffect("Select 1 card, delete 1 card", CanUseCondition, card);
+                activate_class.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, 1, true, EffectDescription());
```
missing activateClass.SetHashstring, On Play, When Digivolving, and when Attacking should have the same SetHashString

S-E-I-G-E replied:
also corrected typo in second WhenAttacking effect description

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_017.cs:
```diff
@@ -0,0 +1,265 @@
+                {
+                    ActivateClass activateClass = new ActivateClass();
+                    activateClass.SetUpICardEffect("Play a token", CanUseCondition, card);
+                    activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, -1, true, EffectDiscription());
```
effect is not optional, change true to false

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_017.cs:
```diff
@@ -0,0 +1,265 @@
+                        return false;
+                    }
+
+                    //#TODO Implement token effect script
```
remove TODO

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_017.cs:
```diff
@@ -0,0 +1,265 @@
+                {
+                    ActivateClass activateClass = new ActivateClass();
+                    activateClass.SetUpICardEffect("Play token", CanUseCondition, card);
+                    activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, -1, true, EffectDiscription());
```
effects is not optional

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_017.cs:
```diff
@@ -0,0 +1,265 @@
+            {
+                ActivateClass activateClass = new ActivateClass();
+                activateClass.SetUpICardEffect("Play token", CanUseCondition, card);
+                activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, 1, true, EffectDiscription());
```
effect is not optional

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_017.cs:
```diff
@@ -0,0 +1,265 @@
+                            canEndNotMax: false,
+                            selectPermanentCoroutine: null,
+                            afterSelectPermanentCoroutine: AfterSelectPermanentCoroutine,
+                            mode: SelectPermanentEffect.Mode.Custom,
```
change mode to .Destory, no need for after select coroutine

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_017.cs:
```diff
@@ -0,0 +1,265 @@
+
+                        yield return ContinuousController.instance.StartCoroutine(new DestroyPermanentsClass(
+                            selectedPermanents,
+                            CardEffectCommons.CardEffectHashtable(activateClass)).Destroy());
```
these are not needed when you change the mode of permanent selection

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_014.cs:
```diff
@@ -0,0 +1,304 @@
+                {
+                    #region On Play
+                    ActivateClass activateClass = new ActivateClass();
+                    activateClass.SetUpICardEffect("", CanUseCondition, card);
```
need short hand effect name

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_014.cs:
```diff
@@ -0,0 +1,304 @@
+                    #region On Play
+                    ActivateClass activateClass = new ActivateClass();
+                    activateClass.SetUpICardEffect("", CanUseCondition, card);
+                    activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, -1, true, EffectDiscription());
```
effect is not optional

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_014.cs:
```diff
@@ -0,0 +1,304 @@
+
+                    bool CanActivateCondition(Hashtable hashtable)
+                    {
+                        return card.Owner.Enemy.GetBattleAreaDigimons().Count(CanSelectPermanent) >= 1;
```
activation conditions are:
- does this card exist in battle area as a digimon
- does your opponent have a digimon meeting the conditions (HasMatchPermanentCondition)

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_014.cs:
```diff
@@ -0,0 +1,304 @@
+                        return card.Owner.Enemy.GetBattleAreaDigimons().Count(CanSelectPermanent) >= 1;
+                    }
+
+                    bool CanSelectPermanent(Permanent permanent)
+                    {
+                        return permanent.DP <= 5000;
+                    }
```
When fixing the activate condtions you will change this to include checking it exists on opponents board as a digimon

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_014.cs:
```diff
@@ -0,0 +1,304 @@
+                    #endregion
+                }
+                {
+                    #region On Digivolve
```
same issues as On Play

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_014.cs:
```diff
@@ -0,0 +1,304 @@
+            {
+                #region End of Turn
+                ActivateClass activateClass = new ActivateClass();
+                activateClass.SetUpICardEffect("", CanUseCondition, card);
```
need short hand effect

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_014.cs:
```diff
@@ -0,0 +1,304 @@
+
+                bool CanUseCondition(Hashtable hashtable)
+                {
+                    return true;
```
use conditions are:
- this card exist on field as a digimon
- and it is your turn

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_014.cs:
```diff
@@ -0,0 +1,304 @@
+                        return true;
+                    }
+                    return false;
+                }
```
the activate conditions are:
- this card exist on field as digimon
- you have another digimon that can suspend for cost
- you have a jesmon in hand that ca be "played" on top of this card

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_014.cs:
```diff
@@ -0,0 +1,304 @@
+                                    isTapped: false,
+                                    root: SelectCardEffect.Root.Hand,
+                                    activateETB: true));
+                                #endregion
```
your not playing a card your using digivolve from hand or trash

S-E-I-G-E replied:
Reworked this based on other examples. 
Select effects now have the layer select a card to suspend, then the DigivolveIntoHandOrTrashCard restricts the digivolve target to Permanent of this card, and restricts the digivolution card from hand to constain [Jesmon]

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_014.cs:
```diff
@@ -0,0 +1,304 @@
+                activateClass.SetIsInheritedEffect(true);
+                cardEffects.Add(activateClass);
+
+                cardEffects.Add(CardEffectFactory.AllianceSelfEffect(false, card, null));
```
this doesn't belong here

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_014.cs:
```diff
@@ -0,0 +1,304 @@
+                {
+                    yield return null;
+                }
+                #endregion
```
this whole effect is wrong. See ST12-013 for reference. Just replace Reeboot for alliance

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_013.cs:
```diff
@@ -0,0 +1,248 @@
+                #region Main Once Per Turn
+                ActivateClass activateClass = new ActivateClass();
+                activateClass.SetUpICardEffect("Play a Digimon with the cost reduced by 2", CanUseCondition, card);
+                activateClass.SetUpActivateClass(null, ActivateCoroutine, 1, false, EffectDiscription());
```
need a set hash string

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_013.cs:
```diff
@@ -0,0 +1,248 @@
+                {
+                    if (CardEffectCommons.IsExistOnBattleArea(card))
+                    {
+                        if (CardEffectCommons.CanActivateSuspendCostEffect(card))
```
nothing suspends for cost here, remove it

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_013.cs:
```diff
@@ -0,0 +1,248 @@
+
+                bool CanUseCondition (Hashtable hashtable)
+                {
+                    if (CardEffectCommons.IsExistOnBattleArea(card))
```
use IsExistOnBattleAreaDigimon instead

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_013.cs:
```diff
@@ -0,0 +1,248 @@
+                    {
+                        return true;
+                    }
+                    return false;
```
you should also check if they can be played as a new permanent

S-E-I-G-E replied:
I think this means checking for open spaces on the board?

S-E-I-G-E replied:
Nevermind, I found an example of CardEffectCommons.CanPlayAsNewPermanent

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_013.cs:
```diff
@@ -0,0 +1,248 @@
+                    return "[Main] [Once Per Turn] You may play 1 Digimon card with [Sistermon] or [Gankoomon] in its name from you hand with the play cost reduced by 2.";
+                }
+
+                bool CanUseCondition (Hashtable hashtable)
```
you should also check if you have a valid target in hand to play

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_013.cs:
```diff
@@ -0,0 +1,248 @@
+
+                            int maxCount = 1;
+
+                            SelectCardEffect selectCardEffect = GManager.instance.GetComponent<SelectCardEffect>();
```
needs to be SelectHandEffect not SelectCardEffect

Michael8818 commented on Assets/CardEffect/BT20/Red/BT20_013.cs:
```diff
@@ -0,0 +1,248 @@
+                #endregion
+            }
+
+            return cardEffects;
```
This is wrong there are plenty of cards that implement this exact thing, look at them ST1-12 does this.

S-E-I-G-E committed:
Revert "Applied feedback"

This reverts commit 6d1ca1bf86aba67de6691222d0f0fa5a5b208238.

S-E-I-G-E committed:
Reverted adding inherited effect default paramteter to Pierce effect

S-E-I-G-E committed:
Merge branch 'workbranch' of https://github.com/DCGO2/DCGO into workbranch

S-E-I-G-E committed:
undo change to pierce

I bungled this process, but I am 99% sure I got it back on track. I have backups of everything in case I need to fix anything from the last 4 commits or so

S-E-I-G-E committed:
undo change to pierce

S-E-I-G-E committed:
Merge branch 'workbranch' of https://github.com/DCGO2/DCGO into workbranch

S-E-I-G-E committed:
Reapply "Applied feedback"

This reverts commit 8232d04fdd52d7efb284ad37feecb968ab791656.

S-E-I-G-E committed:
Applied Second round of feedback. There is one issue I am not 100% sure I resolved correctly. I left it unresolved in the github conversation and commented on it.
